### PR TITLE
Succesfull running of e2e tests part3 on GPU

### PR DESCRIPTION
**Pull request by McHaillet**
Overview of changes:

- Added detection of cuda capable device at the start of full pipeline test to determine whether it can be executed on a GPU. NOTE: I think the desired behavior is that the test is run on both cpu and gpu if possible.
- Final result of test checks whether the unbinned reconstruction reaches < 11 A
- Removed duplicate of generateAlignedTiltImages.py and updated calls of script accordingly
- Switched all the tomogram and subtomogram reconstruction in the test to also use GPU acceleration.
- In reponse to the switch to tomogram reconstruction on GPU, I noticed differences in the exact filter algorithm (partly my fault, as I messed around with it at some point). I now made it working and consistent between CPU/GPU. The default behavior is preserved by setting sliceWidth to 1 if no particle diameter is provided (which was how it was always executed on CPU). In that case its assumed that the particle diameter is equal to the image size (sliceWidth = 1). However the nice thing about this filter is that it adapts for different particle diameters, this functionality is now supported.
- Updated the rotation center for the template in GPU template matching to (sx // 2) to match CPU algorithm. Also switched the interpolation type to 'filt_bspline' as its more accurate. Only with these changes is the behavior almost identical.

### CODE REVIEW

File: pytom/agnostic/filter.py [hunk 1]
```diff
@@ -817,47 +817,25 @@ def exact_filter(tilt_angles, tiltAngle, sX, sY, sliceWidth=1, arr=[]):
...
+    # where frequency = arrCrowther, and 1 / overlap_frequency = sampling/sliceWidth
+    wfuncCrowther = 1. / xp.clip(1 - ((sampling / sliceWidth)[:, xp.newaxis] * arrCrowther) ** 2, 0, 2).sum(axis=0)
 
-    # Create full with weightFunc
+    # Create full width weightFunc
     wfunc = xp.ones((sX, sY), dtype=xp.float32)
 
     # row_stack is not implemented in cupy
-    weightingFunc = xp.column_stack([(wfuncCrowther), ] * (sY))
+    weightingFunc = xp.column_stack((wfuncCrowther, ) * sY)
```
Comment (sroet) [inline pytom/agnostic/filter.py]: In all my testing (numpy only) the following was at least 5x as fast (assuming wfuncCrowther is a 1D array):
```suggestion
    weightingFunc = xp.tile(wfuncCrowther, (sY, 1)).T
```

File: pytom/basic/filter.py [hunk 2]
```diff
@@ -372,45 +372,35 @@ def exactFilter(tilt_angles, tiltAngle, sX, sY, sliceWidth, arr=[]):
...
-    wfuncCrowther = 1. / (clip(1 - array(matrix(abs(sin(diffAngles))).T * arrCrowther)**2, 0, 2)).sum(axis=0)
+    # as in the paper: 1 - frequency / overlap_frequency
+    # where frequency = arrCrowther, and 1 / overlap_frequency = sampling/sliceWidth
+    wfuncCrowther = 1. / np.clip(1 - ((sampling / sliceWidth)[:, np.newaxis] * arrCrowther) ** 2, 0, 2).sum(axis=0)
 
     # Create full with weightFunc
-    wfunc = ones((sX, sY, 1), dtype=float32)
-    wfunc[sX//2-crowtherFreq:sX//2+min(sX//2,crowtherFreq+1),:, 0] = column_stack(([(wfuncCrowther), ] * (sY))).astype(float32)
-
+    wfunc = np.ones((sX, sY, 1), dtype=np.float32)
```
Comment (sroet) [inline pytom/basic/filter.py]: if you are casting to float anyway on line 403, why not set it here already?
```suggestion
    wfunc = np.ones((sX, sY, 1), dtype=float)
```

File: pytom/basic/filter.py [hunk 3]
```diff
@@ -372,45 +372,35 @@ def exactFilter(tilt_angles, tiltAngle, sX, sY, sliceWidth, arr=[]):
...
+    wfunc[sX//2-crowtherFreq:sX//2+min(sX//2,crowtherFreq+1),:, 0] = np.column_stack((wfuncCrowther, ) * sY)
```
Comment (sroet) [inline pytom/basic/filter.py]: In all my testing (numpy only) the following was at least 5x as fast (assuming wfuncCrowther is a 1D array):
```suggestion
    wfunc[sX//2-crowtherFreq:sX//2+min(sX//2,crowtherFreq+1),:, 0] = np.tile(wfuncCrowther, (sY, 1)).T
```

File: pytom/gpu/gpuStructures.py [hunk 4]
```diff
@@ -121,6 +121,8 @@ def template_matching_gpu(self):
         CX, CY, CZ = SX // 2, SY // 2, SZ // 2
         cx, cy, cz = sx // 2, sy // 2, sz // 2
         mx, my, mz = sx % 2, sy % 2, sz % 2
+        # rotation center needs to be set, voltools uses (s - 1) / 2 by default, but pytom_volume s // 2 - s % 2
+        rotation_center = (cx + mx, cy + my, cz + mz)
```
Comment (sroet) [inline pytom/gpu/gpuStructures.py]: either the comment is wrong or the sign on these:
```suggestion
        # rotation center needs to be set, voltools uses (s - 1) / 2 by default, but pytom_volume s // 2 + s % 2
        rotation_center = (cx + mx, cy + my, cz + mz)
```
or
```suggestion
        # rotation center needs to be set, voltools uses (s - 1) / 2 by default, but pytom_volume s // 2 - s % 2
        rotation_center = (cx - mx, cy - my, cz - mz)
```

File: pytom/reconstruction/reconstructionStructures.py [hunk 5]
```diff
@@ -1491,7 +1491,7 @@ def saveParticleProjections(self, particles, projectionSize, binning=1,
 
     # ====== new general to_projection_stack functions ===================
     def to_projection_stack(self, weighting=0, binning=1, low_pass_freq=0.9, apply_circle_filter=True,
-                            scale_factor_particle=1., angle_specimen=0., particle_diameter=280, pixel_size=2.62,
+                            scale_factor_particle=1., angle_specimen=0., particle_diameter=0, pixel_size=2.62,
```
Comment (sroet) [inline pytom/reconstruction/reconstructionStructures.py]: For overwritten defaults I prefer `None`
```suggestion
                            scale_factor_particle=1., angle_specimen=0., particle_diameter=None, pixel_size=2.62,
```

File: pytom/reconstruction/reconstructionStructures.py [hunk 6]
```diff
@@ -1568,9 +1568,10 @@ def to_projection_stack(self, weighting=0, binning=1, low_pass_freq=0.9, apply_c
         offsetStack = vol(1, 2, len(tilts_in_range))
         offsetStack.setAll(0.0)
 
-        # design filters
-        # sliceWidth = imdim  # this should be relative to the particle size as that determines the crowther freq
-        slice_width = (pixel_size * binning * imdim) / particle_diameter
+        if particle_diameter == 0:  # if no particle diameter provided we set slice width relative to full image size
+            slice_width = 1
```
Comment (sroet) [inline pytom/reconstruction/reconstructionStructures.py]:
```suggestion
        if particle_diameter is None:  # if no particle diameter provided we set slice width relative to full image size
            slice_width = 1
        elif particle_diameter <= 0: # Particle diameter should be bigger than 0
            raise ValueError("particle diameter must be greater than 0")
```

File: pytom/reconstruction/reconstructionStructures.py [hunk 7]
```diff
@@ -1690,7 +1691,7 @@ def _retrieve_shared_stack(self, stack_vol):
         stack_vol.copyVolume(npy2vol(copy, len(copy.shape)))
 
     def to_projection_stack_parallel(self, weighting=0, binning=1, low_pass_freq=0.9, apply_circle_filter=True,
-                                     scale_factor_particle=1., angle_specimen=0., particle_diameter=280,
+                                     scale_factor_particle=1., angle_specimen=0., particle_diameter=0,
```
Comment (sroet) [inline pytom/reconstruction/reconstructionStructures.py]: same as above
```suggestion
                                     scale_factor_particle=1., angle_specimen=0., particle_diameter=None,
```

File: pytom/reconstruction/reconstructionStructures.py [hunk 8]
```diff
@@ -1745,10 +1746,10 @@ def to_projection_stack_parallel(self, weighting=0, binning=1, low_pass_freq=0.9
 
         imdim = max(imdimX, imdimY)
 
-        # design filters
-        # sliceWidth = imdim  # this should be relative to the particle size as that determines the crowther freq
-        # set slice width for exact filter
-        slice_width = (pixel_size * binning * imdim) / particle_diameter
+        if particle_diameter == 0:  # if no particle diameter provided we set slice width relative to full image size
+            slice_width = 1
```
Comment (sroet) [inline pytom/reconstruction/reconstructionStructures.py]:
```suggestion
        if particle_diameter is None:  # if no particle diameter provided we set slice width relative to full image size
            slice_width = 1
        elif particle_diameter <= 0: # Particle diameter should be bigger than 0
            raise ValueError("particle diameter must be greater than 0")
```

File: pytom/reconstruction/reconstructionStructures.py [hunk 9]
```diff
@@ -1803,7 +1804,7 @@ def to_projection_stack_parallel(self, weighting=0, binning=1, low_pass_freq=0.9
         return [stack, phiStack, thetaStack, offsetStack]
 
     def to_projection_stack_gpu(self, weighting=0, binning=1, low_pass_freq=0.9, apply_circle_filter=True,
-                                scale_factor_particle=1., angle_specimen=0., particle_diameter=270, pixel_size=2.62,
+                                scale_factor_particle=1., angle_specimen=0., particle_diameter=0, pixel_size=2.62,
```
Comment (sroet) [inline pytom/reconstruction/reconstructionStructures.py]: same as above
```suggestion
                                scale_factor_particle=1., angle_specimen=0., particle_diameter=None, pixel_size=2.62,
```

File: pytom/reconstruction/reconstructionStructures.py [hunk 10]
```diff
@@ -1846,11 +1847,10 @@ def to_projection_stack_gpu(self, weighting=0, binning=1, low_pass_freq=0.9, app
         thetaStack = xp.zeros((len(tilts_in_range)), dtype=xp.float32)
         offsetStack = xp.zeros((len(tilts_in_range), 2), dtype=xp.int32)
 
-        # set slice width for exact filter, smaller objects will have a higher overlap frequency
-        # cut off is calculated as: cut == f_ij / Ny * (dim // 2)
-        # f_ij = 1 / (D * sin th_ij)   => the max overlap frequency of projection i and j
-        # ny = 1 / (2 * px)
-        slice_width = particle_diameter / (pixel_size * binning)
+        if particle_diameter == 0: # if no particle diameter provided we set slice width relative to full image size
+            slice_width = 1
```
Comment (sroet) [inline pytom/reconstruction/reconstructionStructures.py]:
```suggestion
        if particle_diameter is None: # if no particle diameter provided we set slice width relative to full image size
            slice_width = 1
        elif particle_diameter <= 0: # Particle diameter should be bigger than 0
            raise ValueError("particle diameter must be greater than 0")
```

File: tests/E2ETests/test_pytom_FullPipeline.py [hunk 11]
```diff
@@ -543,7 +613,7 @@ def test_18_GLocal_Reduced_Binned_GPU(self):
 
         if not os.path.exists(outdir): os.mkdir(outdir)
 
-        cmd = f'cd {self.projectname}/05_Subtomogram_Analysis; mpiexec -n 5 pytom {self.pytomDir}/bin/GLocalJob.py '
+        cmd = f'cd {self.projectname}/05_Subtomogram_Analysis; mpiexec -n 2 pytom {self.pytomDir}/bin/GLocalJob.py '
```
Comment (sroet) [inline tests/E2ETests/test_pytom_FullPipeline.py]: Shouldn't this be fixed as well?
```suggestion
        cmd = f'cd {self.projectname}/05_Subtomogram_Analysis; mpiexec -n 2 GLocalJob.py '
```

File: tests/E2ETests/test_pytom_FullPipeline.py [hunk 12]
```diff
@@ -680,20 +750,20 @@ def test_20_GLocal_Reduced_NonBinned_GPU(self):
         write(reference, r)
 
 
-        cmd = f'cd {self.projectname}/05_Subtomogram_Analysis; mpiexec -n 5 pytom {self.pytomDir}/bin/GLocalJob.py '
+        cmd = f'cd {self.projectname}/05_Subtomogram_Analysis; mpiexec -n 2 pytom {self.pytomDir}/bin/GLocalJob.py '
```
Comment (sroet) [inline tests/E2ETests/test_pytom_FullPipeline.py]: same as above
```suggestion
        cmd = f'cd {self.projectname}/05_Subtomogram_Analysis; mpiexec -n 2 GLocalJob.py '
```

**Review by sroet — CHANGES_REQUESTED** [review]
Couple of comments, Looks good otherwise

**Comment by McHaillet** [comment]
The part of the tests for which the code was updated passed for me!

**Comment by sroet** [comment]
> The part of the tests for which the code was updated passed for me!

I will merge then :)

**Pull request merged**